distributed/joblib.py

```diff
@@ -130,17 +157,40 @@ def callback_wrapper():
             if callback is not None:
                 callback(result)  # gets called in separate thread
 
-        self.client.loop.add_callback(callback_wrapper)
+        client.loop.add_callback(callback_wrapper)
 
         future.get = future.result  # monkey patch to achieve AsyncResult API
         return future
 
     def abort_everything(self, ensure_ready=True):
         # Tell the client to cancel any task submitted via this instance
         # as joblib.Parallel will never access those results.
-        self.client.cancel(self.futures)
+        self.get_client().cancel(self.futures)
         self.futures.clear()
 
+    @contextlib.contextmanager
+    def retrieval_context(self):
+        """Override ParallelBackendBase.retrieval_context to avoid deadlocks.
+
+        This removes thread from the worker's thread pool (using 'secede'), and
+        rejoins when finished (using 'rejoin'). This allows nested parallelism
+        using joblib without deadlock.
+        """
+        # See 'joblib.Parallel.__call__' and 'joblib.Parallel.retrieve' for how
+        # this is used.
+
+        try:
+            secede()
+        except ValueError:
+            pass
+
+        yield
+
+        try:
+            rejoin()
+        except AttributeError:
+            pass
```

Comment:
Should we rejoin here?  Typically this is necessary if there is additional computational work to do in the task.  That may not be the case here.  It's harmless either way, but might add a tiny bit of unwanted delay.

Reply:
I wasn't sure about that either. To make sure I understand, when the worker thread `secedes`, the scheduler will tell the worker to make a new thread to take its place? In that case, yes we should be OK just not rejoining.

Reply:
Right, the thread leaves the thread pool, leaving it with `n - 1` threads.  We make a new thread to take it's place.  Relevant code here: https://github.com/dask/distributed/blob/master/distributed/threadpoolexecutor.py#L94-L105

```python
def secede(adjust=True):
    """ Have this thread secede from the ThreadPoolExecutor
    See Also
    --------
    rejoin: rejoin the thread pool
    """
    thread_state.proceed = False
    with threads_lock:
        thread_state.executor._threads.remove(threading.current_thread())
        if adjust:
            thread_state.executor._adjust_thread_count()
```